### PR TITLE
Kernel: Improve threading & scheduling V3

### LINKED ISSUE #6538: [Dead or Alive Xtreme 3 Scarlet] Stuck on launching screen if Multicore is turned on.
-The game will be stuck on the launching screen on versions of Yuzu after 570.
-The solution is to **turn off the Multicore option** at the bottom of the screen, and the game will boot and run successfully.
-Would appreciate if the team could look into this. Thank you.

### PR DESCRIPTION
**Pull request by bunnei**
The primary goal of this PR is to continue to improve the accuracy of our scheduler and threading implementations, improve parity with recent HOS updates, and fix race conditions/deadlocks/crashes.

There are several changes here:
* Implement more complete tracking of KThread dispatch count through KScopedDisableDispatch.
* Improved KScheduler EnableScheduling/DisableScheduling routines as a result.
* Overall simplification of a few KScheduler implementations.
* Improved emulation of host-only HLE threads as simulated "guest threads" on core 3.
* Use std::jthread for CPU threads.
* Use std::jthread for service host threads. This fixes a deadlock that can cause a crash in Pokémon Sword & Shield.
* Improvements to thread pinning code.
* Updated our KThreadQueue implementation, and updated our synchronization primitives to use the new pattern.
* Fixed several kernel shutdown crashes.
* Updated KConditionVariable, KAddressArbiter, KLightLock, KSynchronizationObject, etc. implementations.

Fixes #6538

Supersedes #7398 and #6964 .

Credit to @FernandoS27 for helping with a few fixes with this change, and thank you to @gidoly, @G-Spawn, and others for all of the help testing.

### CODE REVIEW

File: src/core/hle/kernel/k_condition_variable.cpp [hunk 1]
```diff
@@ -189,35 +223,24 @@ KThread* KConditionVariable::SignalImpl(KThread* thread) {
...
-    // Prepare for signaling.
-    constexpr int MaxThreads = 16;
-
-    KLinkedList<KThread> thread_list{kernel};
-    std::array<KThread*, MaxThreads> thread_array;
-    s32 num_to_close{};
-
     // Perform signaling.
-    s32 num_waiters{};
+    int num_waiters = 0;
```
Comment (lioncash) [inline src/core/hle/kernel/k_condition_variable.cpp:243]: Why were these changed back to `= 0`? I thought `{}` was the preferred way to initialize variables to defaults now

File: src/core/hle/kernel/k_condition_variable.cpp [hunk 2]
```diff
@@ -226,46 +249,28 @@ void KConditionVariable::Signal(u64 cv_key, s32 count) {
...
+            this->SignalImpl(target_thread);
             it = thread_tree.erase(it);
             target_thread->ClearConditionVariable();
             ++num_waiters;
         }
 
         // If we have no waiters, clear the has waiter flag.
         if (it == thread_tree.end() || it->GetConditionVariableKey() != cv_key) {
-            const u32 has_waiter_flag{};
+            const u32 has_waiter_flag = 0;
```
Comment (lioncash) [inline src/core/hle/kernel/k_condition_variable.cpp:260]: Ditto

File: src/core/hle/kernel/k_light_condition_variable.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace {
+
+class ThreadQueueImplForKLightConditionVariable final : public KThreadQueue {
+public:
+    ThreadQueueImplForKLightConditionVariable(KernelCore& kernel_, KThread::WaiterList* wl,
+                                              bool term)
+        : KThreadQueue(kernel_), m_wait_list(wl), m_allow_terminating_thread(term) {}
+
+    virtual void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.cpp:21]:
```suggestion
    void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 4]
```diff
@@ -1,73 +1,25 @@
-// Copyright 2020 yuzu Emulator Project
+// Copyright 2021 yuzu Emulator Project
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:1]: This doesn't need to be updated, it's just to establish a creation year

File: src/core/hle/kernel/k_light_lock.cpp [hunk 5]
```diff
@@ -5,98 +5,100 @@
...
 
 namespace Kernel {
 
+namespace {
+
+class ThreadQueueImplForKLightLock final : public KThreadQueue {
+public:
+    explicit ThreadQueueImplForKLightLock(KernelCore& kernel_) : KThreadQueue(kernel_) {}
+
+    virtual void CancelWait([[maybe_unused]] KThread* waiting_thread,
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_lock.cpp:19]:
```suggestion
    void CancelWait([[maybe_unused]] KThread* waiting_thread,
```

File: src/core/hle/kernel/k_process.h [hunk 6]
```diff
@@ -259,7 +259,7 @@ class KProcess final
 
     [[nodiscard]] KThread* GetPinnedThread(s32 core_id) const {
         ASSERT(0 <= core_id && core_id < static_cast<s32>(Core::Hardware::NUM_CPU_CORES));
-        return pinned_threads[core_id];
+        return pinned_threads.at(core_id);
```
Comment (lioncash) [inline src/core/hle/kernel/k_process.h:262]: Doesn't the assertion already sanitize the inputs?

File: src/core/hle/kernel/k_synchronization_object.cpp [hunk 7]
```diff
@@ -8,11 +8,66 @@
...
 
+namespace {
+
+class ThreadQueueImplForKSynchronizationObjectWait final : public KThreadQueueWithoutEndWait {
+public:
+    ThreadQueueImplForKSynchronizationObjectWait(KernelCore& kernel_, KSynchronizationObject** o,
+                                                 KSynchronizationObject::ThreadListNode* n, s32 c)
+        : KThreadQueueWithoutEndWait(kernel_), m_objects(o), m_nodes(n), m_count(c) {}
+
+    virtual void NotifyAvailable(KThread* waiting_thread, KSynchronizationObject* signaled_object,
```
Comment (lioncash) [inline src/core/hle/kernel/k_synchronization_object.cpp:25]:
```suggestion
    void NotifyAvailable(KThread* waiting_thread, KSynchronizationObject* signaled_object,
```

File: src/core/hle/kernel/k_synchronization_object.cpp [hunk 8]
```diff
@@ -8,11 +8,66 @@
...
+        waiting_thread->SetSyncedIndex(sync_index);
+
+        // Set the waiting thread as not cancellable.
+        waiting_thread->ClearCancellable();
+
+        // Invoke the base end wait handler.
+        KThreadQueue::EndWait(waiting_thread, wait_result);
+    }
+
+    virtual void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```
Comment (lioncash) [inline src/core/hle/kernel/k_synchronization_object.cpp:49]:
```suggestion
    void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```

File: src/core/hle/kernel/k_thread.cpp [hunk 9]
```diff
@@ -56,6 +59,34 @@ static void ResetThreadContext64(Core::ARM_Interface::ThreadContext64& context,
...
+    explicit ThreadQueueImplForKThreadSleep(KernelCore& kernel_)
+        : KThreadQueueWithoutEndWait(kernel_) {}
+};
+
+class ThreadQueueImplForKThreadSetProperty final : public KThreadQueue {
+public:
+    explicit ThreadQueueImplForKThreadSetProperty(KernelCore& kernel_, KThread::WaiterList* wl)
+        : KThreadQueue(kernel_), m_wait_list(wl) {}
+
+    virtual void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread.cpp:75]:
```suggestion
    void CancelWait(KThread* waiting_thread, ResultCode wait_result,
```

File: src/core/hle/kernel/k_thread_queue.h [hunk 10]
```diff
@@ -4,78 +4,32 @@
...
 private:
     KernelCore& kernel;
     KThread::WaiterList wait_list{};
 };
 
+class KThreadQueueWithoutEndWait : public KThreadQueue {
+public:
+    explicit KThreadQueueWithoutEndWait(KernelCore& kernel_) : KThreadQueue(kernel_) {}
+
+    virtual void EndWait(KThread* waiting_thread, ResultCode wait_result) override final;
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread_queue.h:32]:
```suggestion
    void EndWait(KThread* waiting_thread, ResultCode wait_result) override final;
```

File: src/core/hle/kernel/k_process.h [hunk 11]
```diff
@@ -368,14 +369,14 @@ class KProcess final
     void PinThread(s32 core_id, KThread* thread) {
         ASSERT(0 <= core_id && core_id < static_cast<s32>(Core::Hardware::NUM_CPU_CORES));
         ASSERT(thread != nullptr);
-        ASSERT(pinned_threads[core_id] == nullptr);
+        ASSERT(pinned_threads.at(core_id) == nullptr);
```
Comment (lioncash) [inline src/core/hle/kernel/k_process.h:372]: Ditto wrt the assertion above already sanitizing the range and the one below in UnpinThread

File: src/core/hle/kernel/k_scheduler.cpp [hunk 12]
```diff
@@ -376,20 +376,28 @@ void KScheduler::ClearSchedulerUpdateNeeded(KernelCore& kernel) {
...
-        ASSERT(scheduler->GetCurrentThread()->GetDisableDispatchCount() >= 1);
-        if (scheduler->GetCurrentThread()->GetDisableDispatchCount() >= 1) {
-            scheduler->GetCurrentThread()->EnableDispatch();
-        }
+    // If we are shutting down the kernel, none of this is relevant anymore.
+    if (kernel.IsShuttingDown()) {
+        return;
+    }
+
+    ASSERT(GetCurrentThreadPointer(kernel)->GetDisableDispatchCount() >= 1);
```
Comment (lioncash) [inline src/core/hle/kernel/k_scheduler.cpp:394]: Can't the result of `GetCurrentThreadPointer()` be stored into a local here instead of calling it three times in succession to do work with it?

File: src/core/hle/kernel/k_condition_variable.cpp (hunk 1 above)
Comment (heinermann) [reply src/core/hle/kernel/k_condition_variable.cpp:243]: I feel like `= 0` is more direct, explicit, and faster to read than having to scan the type and recall what it default initializes to.

Comment (vonchenplus) [reply src/core/hle/kernel/k_condition_variable.cpp:243]: Shouldn't we be using s32 instead of int here?

Comment (lioncash) [reply src/core/hle/kernel/k_condition_variable.cpp:243]: @heinermann I don't disagree, however, convention changes are best suited for their own PRs

**Comment by yuwenhua653** [comment]
Doesn't seem to be fixed #5788

**Comment by Oschowa** [comment]
I"ve tested this branch on linux and it seems like all my games deadlock on launch with it. Disabling CPU multicore, Async GPU or switching between Vulkan/GL doesn't help. 
[yuzu_log.txt](https://github.com/yuzu-emu/yuzu/files/7626494/yuzu_log.txt)

**Comment by vonchenplus** [comment]
Seems to cause more frequent crashes #7477

**Comment by FernandoS27** [comment]
It does not help that you guys call evrything a crash. Be more specific, does the emulator close, the game stops working (a game crash) or game softlock? Also which games?

**Comment by LOuroboros** [comment]
> It does not help that you guys call evrything a crash. Be more specific, does the emulator close, the game stops working (a game crash) or game softlock? Also which games?

In the case of #7477 it's Pokémon Brilliant Diamond v1.1.1, and what I call a crash is the action of Yuzu closing itself out of nowhere, completely on its own and without any input from my part, for whatever that's worth.

That has me, playing the game normally, and then Yuzu closing itself without me doing anything except play the game normally.

**Comment by Oschowa** [comment]
To clarify my earlier comment, i"m seeing softlock on linux when launching games (Yuzu basically hangs on the "launching game" screen) with this PR, while it works with latest master and mainline appimage. It happens with all game i have, but specifically with Tokyo Mirage Sessions, The World Ends With You Final Remix and SMTV.

**Comment by FernandoS27** [comment]
I'll run this through Valgrind later.

**Comment by dietolive-tw** [comment]
The deadlock on linux is caused by missing stop_requested while ServiceThread::Impl::~Impl() being called. Keeping the previous code and send stop_requested manually can solve this, but i don't really know if this is the correct method to do, I am not familiar in c++.

ServiceThread::Impl::~Impl() {
    condition.notify_all();
    for (std::jthread& thread : threads) {
        thread.request_stop();
        thread.join();
    }
}

**Comment by bunnei** [comment]
Re-tagging, as I believe the previously reported issues to be fixed. Please keep an eye out, especially with BDSP.

If you experience a regression, please also diligently try to reproduce it with mainline.

**Comment by yuwenhua653** [comment]
In EA2286 Pokémon Sword and Shield automatic archive is still soft lock

**Comment by bunnei** [comment]
> In EA2286 Pokémon Sword and Shield automatic archive is still soft lock

This is not a regression with this PR though, right? (can you confirm this happens in mainline)

**Comment by yuwenhua653** [comment]
> 



> > 在EA2286宝可梦剑盾自动存档还是软锁
> 
> 不过，这不是这个 PR 的回归，对吧？（你能确认这发生在主线中吗）



> > In EA2286 Pokémon Sword and Shield automatic archive is still soft lock
> 
> This is not a regression with this PR though, right? (can you confirm this happens in mainline)

Yes, this problem has been existed

**Comment by yuwenhua653** [comment]
> > In EA2286 Pokémon Sword and Shield automatic archive is still soft lock
> 
> This is not a regression with this PR though, right? (can you confirm this happens in mainline)

Yes, this problem has been existed

**Comment by FernandoS27** [comment]
I think the wording of "pokemon sword crashes" was too wide and generated confusion. It mostly fixed the "save ones" on DLC.

**Review by lioncash — APPROVED** [review]

**Review by Morph1984 — APPROVED** [review]

**Comment by yuwenhua653** [comment]
> I think the wording of "pokemon sword crashes" was too wide and generated confusion. It mostly fixed the "save ones" on DLC.

No, DLC SAVE still has problems

**Comment by gandalfthewhite19890404** [comment]
Test, please, these issues with PR:
https://github.com/yuzu-emu/yuzu/issues/6058
https://github.com/yuzu-emu/yuzu/issues/6368

**Comment by SpecialDarkwolf999** [comment]
diablo 3 works fine on EA 2305 now

**Comment by elkay** [comment]
Dead or Alive Scarlet is booting again on EA 2305 with multicore on.
